### DISCUSSION TITLE
Returning a reference

### DISCUSSION THREAD

**Discussion by devongovett** [napi-rs/napi-rs#1142] [Q&A] [answered]
I'm looking into implementing the [CSS Typed OM](https://www.w3.org/TR/css-typed-om-1/) API for [Parcel CSS](https://github.com/parcel-bundler/parcel-css). This has a complex hierarchy of objects, with fields that include other objects, etc. One thing I ran into is that it currently doesn't seem easy (or possible?) to have a getter or method return the same instance of an object each time it is called.

I tried returning a reference, but this seems to currently panic:

https://github.com/napi-rs/napi-rs/blob/4f422b6625daaddf75bb422331e0b80c14c9b901/crates/backend/src/codegen/struct.rs#L102-L112

I saw the `Reference` API in a recent release, but I'm not sure if this is related or not.

Say I have the following structs:

```rust
#[napi]
struct Foo {
  bar: Bar
}

#[napi]
struct Bar {
  baz: u32
}

#[napi]
impl Foo {
  #[napi(getter)]
  pub fn bar(&self) -> ??? {
    self.bar
  }
}
```

How can I make `bar` return the same instance of `Bar` each time it is called?

**Comment by Brooooooklyn** [comment, marked as answer]
The `2.3.0` released a new feature: [Reference API](https://github.com/napi-rs/napi-rs/releases/tag/napi%402.3.0).
I think it's perfectly suitable for your scenario. I haven't had time to write the documentation for this feature but you can see the real-world example: https://github.com/Brooooooklyn/simple-git/blob/main/src/repo.rs#L211.

@devongovett

**Reply by devongovett** [reply]
Thanks for this example! It looks like I cannot return a `SharedReference` from an napi function though. I get "the trait bound `napi::bindgen_prelude::SharedReference<stylesheet::CSSStyleSheet, stylesheet::CSSRuleList>: NapiRaw` is not satisfied". In your example, it looks like it's returning a new instance of a wrapper object pointing to the inner reference each time. So, running a small example:

```js
import { Repository } from '@napi-rs/simple-git'
const repo = new Repository('/path/to/repo') // Open an existed repo
repo.head() == repo.head() // false
```

Is it possible to somehow make this return true?

**Reply by devongovett** [reply]
Here's one way I came up with, but it's a bit manual and required some unsafe code... 

```rust
struct CSSStyleSheet {
  stylesheet: Rc<RefCell<OwnedStyleSheet>>,
  rules: Option<Ref<()>>,
}

#[napi]
impl CSSStyleSheet {
  #[napi(getter)]
  pub fn css_rules(&mut self, env: Env) -> Result<JsUnknown> {
    if let Some(rules) = &self.rules {
      return env.get_reference_value(rules);
    }

    let rules = CSSRuleList {
      stylesheet: self.stylesheet.clone(),
      rules: Vec::new(),
    };

    let value = unsafe {
      let napi_value = napi::bindgen_prelude::ToNapiValue::to_napi_value(env.raw(), rules)?;
      napi::JsUnknown::from_napi_value(env.raw(), napi_value)?
    };

    self.rules = Some(env.create_reference_with_refcount(&value, 0)?);
    Ok(value)
  }
}
```

It seems to work ok, but maybe something like that could be wrapped up a little nicer.

**Reply by Brooooooklyn** [reply]
Does this meet your requirement? https://github.com/napi-rs/napi-rs/blob/into-reference/examples/napi/src/reference.rs#L60-L101

**Reply by devongovett** [reply]
Yeah that looks really nice. Thank you!
